keylime/ima/ima.py

```diff
@@ -625,46 +462,69 @@ def ima_policy_db_contents(ima_policy_name: str, ima_policy: str, tpm_policy: st
-            else:
-                alist_bytes = json.dumps(copy.deepcopy(EMPTY_ALLOWLIST)).encode()
-                sha256 = hashlib.sha256()
-                sha256.update(alist_bytes)
-                ima_policy_db_format["checksum"] = sha256.hexdigest()
+    if "meta" in ima_policy_json:
+        if "checksum" in ima_policy_json["meta"]:
+            ima_policy_db_format["checksum"] = ima_policy_json["meta"]["checksum"]
+        else:
+            alist_bytes = json.dumps(ima_policy).encode()
+            sha256 = hashlib.sha256()
+            sha256.update(alist_bytes)
+            ima_policy_db_format["checksum"] = sha256.hexdigest()
     if tpm_policy:
         ima_policy_db_format["tpm_policy"] = tpm_policy
     return ima_policy_db_format
 
 
-def read_excllist(exclude_path=None):
-    excl_list = []
-    if exclude_path and os.path.exists(exclude_path):
-        with open(exclude_path, encoding="utf-8") as f:
-            for line in f:
-                line = line.strip()
-                if line.startswith("#") or len(line) == 0:
-                    continue
-                excl_list.append(line)
+def verify_ima_policy(ima_policy, ima_policy_key=None, ima_policy_sig=None, verify_sig=True):
+    """
+    Verify that an IMA policy is valid. If provided IMA policy has a detached signature, verify the signature.
+    """
+
+    if ima_policy is None:
+        raise ImaValidationError("No IMA policy provided!", 400)
+
+    if verify_sig:
+        if not ima_policy_sig:
+            raise ImaValidationError(
+                message="Verifier is enforcing signature validation, but no signature was provided.", code=405
+            )
+        if not ima_policy_key:
+            raise ImaValidationError(
+                message="Verifier is enforcing signature validation, but no public key was provided.", code=405
+            )
+
+        if not signing.verify_signature(ima_policy_key, ima_policy_sig, ima_policy):
+            raise ImaValidationError(message="IMA policy failed detached signature verification!", code=401)
+        logger.info("IMA policy passed detached signature verification")
 
-        logger.debug("Loaded exclusion list from %s: %s", exclude_path, excl_list)
+    # validate that the allowlist is proper JSON
+    try:
+        lists = json.loads(ima_policy)
+    except Exception as error:
+        raise ImaValidationError(message="IMA policy is not valid JSON!", code=400) from error
+
+    # Validate exlude list contains valid regular expressions
+    _, excl_err_msg = validators.valid_exclude_list(lists.get("exclude"))
+    if excl_err_msg:
+        raise ImaValidationError(
+            message=excl_err_msg + " Exclude list regex is misformatted. Please correct the issue and try again.",
+            code=400,
+        )
 
-    return excl_list
+
+def deserialize_ima_policy(ima_policy):
+    """
+    TODO: Extract IMA policy JSON from DSSE envelope, if applicable.
```

Comment:
Is this TODO real? If so add it as a comment in the code, not as a doc-string

Reply:
@aplanas Thanks, fixed - I've also added a proper docstring to that function too.